File: app/routers/loans.py
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session
from typing import List, Optional
from datetime import timedelta
from app.database import get_db
from app.models import Loan, Client, LoanStatus, PaymentFrequency, User
from app.schemas import (
    LoanCreate, LoanOut, LoanDetail, PaymentOut, AmortizationRow,
    SimulateRequest, SimulateResponse,
)
from app.auth import get_current_user
from app.utils.loan_calculator import (
    calculate_periodic_payment,
    generate_amortization_table,
    calculate_total_amount,
)

router = APIRouter(prefix="/api/loans", tags=["Loans"])


def _get_end_date(loan):
    freq_days = {"DAILY": 1, "WEEKLY": 7, "MONTHLY": 30}
    days = freq_days.get(loan.payment_frequency.value if hasattr(loan.payment_frequency, 'value') else loan.payment_frequency, 30)
    return loan.start_date + timedelta(days=days * loan.total_periods)


def _loan_to_out(loan: Loan) -> LoanOut:
    return LoanOut(
        id=loan.id,
        client_id=loan.client_id,
        client_name=loan.client.full_name if loan.client else "",
        principal=loan.principal,
        interest_rate=loan.interest_rate,
        interest_type=loan.interest_type.value if loan.interest_type else "FIXED",
        payment_frequency=loan.payment_frequency.value if loan.payment_frequency else "MONTHLY",
        total_periods=loan.total_periods,
        total_amount=loan.total_amount,
        outstanding_balance=loan.outstanding_balance,
        status=loan.status.value if loan.status else "ACTIVE",
        penalty_rate=loan.penalty_rate,
        start_date=loan.start_date,
        end_date=loan.end_date,
        created_at=loan.created_at,
        payments_count=len(loan.payments) if loan.payments else 0,
    )


@router.get("", response_model=List[LoanOut])
def list_loans(
    status: Optional[str] = Query(None),
    client_id: Optional[int] = Query(None),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    q = db.query(Loan)
    if status:
        q = q.filter(Loan.status == status)
    if client_id:
        q = q.filter(Loan.client_id == client_id)
    loans = q.order_by(Loan.created_at.desc()).all()
    return [_loan_to_out(l) for l in loans]


@router.post("", response_model=LoanOut, status_code=201)
def create_loan(
    req: LoanCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    client = db.query(Client).filter(Client.id == req.client_id).first()
    if not client:
        raise HTTPException(status_code=404, detail="Cliente no encontrado")

    total = calculate_total_amount(req.principal, req.interest_rate, req.total_periods)

    loan = Loan(
        client_id=req.client_id,
        principal=req.principal,
        interest_rate=req.interest_rate,
        interest_type=req.interest_type.value,
        payment_frequency=req.payment_frequency.value,
        total_periods=req.total_periods,
        total_amount=total,
        outstanding_balance=req.principal,
        status=LoanStatus.ACTIVE,
        penalty_rate=req.penalty_rate,
        start_date=req.start_date,
    )
    loan.end_date = _get_end_date(loan)

    db.add(loan)
    db.commit()
    db.refresh(loan)
    return _loan_to_out(loan)


@router.get("/{loan_id}", response_model=LoanDetail)
def get_loan(
    loan_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    loan = db.query(Loan).filter(Loan.id == loan_id).first()
    if not loan:
        raise HTTPException(status_code=404, detail="Préstamo no encontrado")

    table = generate_amortization_table(loan.principal, loan.interest_rate, loan.total_periods)
    amort_rows = [AmortizationRow(**row) for row in table]

    payments_out = []
    for p in loan.payments:
        payments_out.append(PaymentOut(
            id=p.id,
            loan_id=p.loan_id,
            amount=p.amount,
            principal_portion=p.principal_portion,
            interest_portion=p.interest_portion,
            penalty_portion=p.penalty_portion or 0.0,
            balance_after=p.balance_after,
            payment_date=p.payment_date,
            receipt_number=p.receipt_number,
            notes=p.notes,
            created_at=p.created_at,
        ))

    out = _loan_to_out(loan)
    return LoanDetail(
        **out.model_dump(),
        payments=payments_out,
        amortization_table=amort_rows,
    )


@router.put("/{loan_id}/status")
def update_loan_status(
    loan_id: int,
    new_status: str = Query(...),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    loan = db.query(Loan).filter(Loan.id == loan_id).first()
    if not loan:
        raise HTTPException(status_code=404, detail="Préstamo no encontrado")

    try:
        loan.status = LoanStatus(new_status)
    except ValueError:
        raise HTTPException(status_code=400, detail="Estado inválido. Usar: ACTIVE, PAID, DELINQUENT")

    db.commit()
    return {"detail": f"Estado actualizado a {new_status}"}


@router.post("/simulate", response_model=SimulateResponse)
def simulate_loan(req: SimulateRequest):
    periodic = calculate_periodic_payment(req.principal, req.interest_rate, req.total_periods)
    table = generate_amortization_table(req.principal, req.interest_rate, req.total_periods)
    total = round(sum(row["payment"] for row in table), 2)

    return SimulateResponse(
        principal=req.principal,
        interest_rate=req.interest_rate,
        payment_frequency=req.payment_frequency.value,
        total_periods=req.total_periods,
        periodic_payment=periodic,
        total_amount=total,
        total_interest=round(total - req.principal, 2),
        table=[AmortizationRow(**row) for row in table],
    )

File: static/js/loans.js
/* ══════════════════════════════════════════════════
   Loans View + Simulator
   ══════════════════════════════════════════════════ */

async function renderLoans() {
    const main = document.getElementById('main-content');
    main.innerHTML = `
        <div class="fade-in-up">
            <div class="page-header">
                <h1>💳 Préstamos</h1>
                <div class="header-actions">
                    <select id="loan-filter" class="btn btn-secondary" style="padding:0.5rem 1rem;background:var(--bg-input);border:1px solid var(--border);border-radius:var(--radius-sm);color:var(--text-primary);font-family:var(--font);">
                        <option value="">Todos</option>
                        <option value="ACTIVE">Activos</option>
                        <option value="PAID">Saldados</option>
                        <option value="DELINQUENT">En Mora</option>
                    </select>
                    <button class="btn btn-primary" id="btn-new-loan">+ Nuevo Préstamo</button>
                </div>
            </div>
            <div class="card">
                <div class="table-container">
                    <table>
                        <thead>
                            <tr>
                                <th>ID</th>
                                <th>Cliente</th>
                                <th>Capital</th>
                                <th>Tasa</th>
                                <th>Frecuencia</th>
                                <th>Saldo</th>
                                <th>Estado</th>
                                <th>Pagos</th>
                            </tr>
                        </thead>
                        <tbody id="loans-table-body">
                            <tr><td colspan="8" style="text-align:center;padding:2rem;color:var(--text-muted);">Cargando...</td></tr>
                        </tbody>
                    </table>
                </div>
            </div>
        </div>
    `;

    document.getElementById('btn-new-loan').addEventListener('click', () => showNewLoanForm());
    document.getElementById('loan-filter').addEventListener('change', (e) => loadLoans(e.target.value));

    await loadLoans();
}

async function loadLoans(statusFilter = '') {
    try {
        const params = {};
        if (statusFilter) params.status = statusFilter;
        const loans = await API.getLoans(params);
        const tbody = document.getElementById('loans-table-body');

        if (loans.length === 0) {
            tbody.innerHTML = `<tr><td colspan="8"><div class="empty-state"><h3>No hay préstamos</h3><p>Crea un préstamo desde la ficha de un cliente</p></div></td></tr>`;
            return;
        }

        const freqMap = { 'DAILY': 'Diario', 'WEEKLY': 'Semanal', 'MONTHLY': 'Mensual' };

        tbody.innerHTML = loans.map(l => `
            <tr class="clickable" onclick="viewLoan(${l.id})">
                <td>#${l.id}</td>
                <td>${l.client_name}</td>
                <td class="money">${formatMoney(l.principal)}</td>
                <td>${l.interest_rate}%</td>
                <td>${freqMap[l.payment_frequency] || l.payment_frequency}</td>
                <td class="money">${formatMoney(l.outstanding_balance)}</td>
                <td>${statusBadge(l.status)}</td>
                <td>${l.payments_count}</td>
            </tr>
        `).join('');
    } catch (e) {
        showToast('Error cargando préstamos: ' + e.message, 'error');
    }
}

async function showNewLoanForm(clientId = null, clientName = '') {
    let clientsOptions = '';
    try {
        const clients = await API.getClients();
        clientsOptions = clients.map(c =>
            `<option value="${c.id}" ${c.id === clientId ? 'selected' : ''}>${c.full_name} (${c.id_number})</option>`
        ).join('');
    } catch (e) { /* ignore */ }

    const today = new Date().toISOString().split('T')[0];

    showGenericModal(`
        <div class="modal-header">
            <h2>Nuevo Préstamo${clientName ? ` — ${clientName}` : ''}</h2>
            <button class="modal-close" onclick="closeModal('generic-modal')">&times;</button>
        </div>
        <div id="loan-form">
            <div class="form-group">
                <label>Cliente</label>
                <select id="lf-client">${clientsOptions}</select>
            </div>
            <div class="form-row">
                <div class="form-group">
                    <label>Capital ($)</label>
                    <input type="number" id="lf-principal" min="1" step="0.01" placeholder="10000">
                </div>
                <div class="form-group">
                    <label>Tasa de Interés (% por periodo)</label>
                    <input type="number" id="lf-rate" min="0.01" step="0.01" placeholder="5">
                </div>
            </div>
            <div class="form-row">
                <div class="form-group">
                    <label>Frecuencia de Cobro</label>
                    <select id="lf-frequency">
                        <option value="MONTHLY">Mensual</option>
                        <option value="WEEKLY">Semanal</option>
                        <option value="DAILY">Diario</option>
                    </select>
                </div>
                <div class="form-group">
                    <label>Número de Periodos</label>
                    <input type="number" id="lf-periods" min="1" placeholder="12">
                </div>
            </div>
            <div class="form-row">
                <div class="form-group">
                    <label>Fecha de Inicio</label>
                    <input type="date" id="lf-start" value="${today}">
                </div>
                <div class="form-group">
                    <label>Tasa de Mora (%)</label>
                    <input type="number" id="lf-penalty" value="2" min="0" step="0.01">
                </div>
            </div>
            <button type="button" class="btn btn-primary btn-full" id="lf-submit">Crear Préstamo</button>
        </div>
    `);

    document.getElementById('lf-submit').addEventListener('click', async () => {
        try {
            const data = {
                client_id: parseInt(document.getElementById('lf-client').value),
                principal: parseFloat(document.getElementById('lf-principal').value),
                interest_rate: parseFloat(document.getElementById('lf-rate').value),
                interest_type: 'FIXED',
                payment_frequency: document.getElementById('lf-frequency').value,
                total_periods: parseInt(document.getElementById('lf-periods').value),
                penalty_rate: parseFloat(document.getElementById('lf-penalty').value),
                start_date: document.getElementById('lf-start').value,
            };
            await API.createLoan(data);
            showToast('Préstamo creado exitosamente', 'success');
            closeModal('generic-modal');
            if (document.getElementById('loans-table-body')) loadLoans();
        } catch (e) {
            showToast(e.message, 'error');
        }
    });
}

async function viewLoan(id) {
    try {
        const loan = await API.getLoan(id);
        const freqMap = { 'DAILY': 'Diario', 'WEEKLY': 'Semanal', 'MONTHLY': 'Mensual' };

        showGenericModal(`
            <div class="modal-header">
                <h2>Préstamo #${loan.id}</h2>
                <button class="modal-close" onclick="closeModal('generic-modal')">&times;</button>
            </div>
            <div style="display:grid;grid-template-columns:1fr 1fr;gap:1rem;margin-bottom:1.5rem;">
                <div>
                    <span style="color:var(--text-muted);font-size:0.8rem;">Cliente</span>
                    <p style="font-weight:600;">${loan.client_name}</p>
                </div>
                <div>
                    <span style="color:var(--text-muted);font-size:0.8rem;">Estado</span>
                    <p>${statusBadge(loan.status)}</p>
                </div>
                <div>
                    <span style="color:var(--text-muted);font-size:0.8rem;">Capital</span>
                    <p style="font-weight:600;" class="money">${formatMoney(loan.principal)}</p>
                </div>
                <div>
                    <span style="color:var(--text-muted);font-size:0.8rem;">Saldo Pendiente</span>
                    <p style="font-weight:600;color:var(--warning);" class="money">${formatMoney(loan.outstanding_balance)}</p>
                </div>
                <div>
                    <span style="color:var(--text-muted);font-size:0.8rem;">Tasa</span>
                    <p>${loan.interest_rate}% ${freqMap[loan.payment_frequency] || ''}</p>
                </div>
                <div>
                    <span style="color:var(--text-muted);font-size:0.8rem;">Total a Pagar</span>
                    <p class="money">${formatMoney(loan.total_amount)}</p>
                </div>
                <div>
                    <span style="color:var(--text-muted);font-size:0.8rem;">Inicio</span>
                    <p>${formatDate(loan.start_date)}</p>
                </div>
                <div>
                    <span style="color:var(--text-muted);font-size:0.8rem;">Vencimiento</span>
                    <p>${formatDate(loan.end_date)}</p>
                </div>
            </div>

            <div style="display:flex;gap:0.5rem;margin-bottom:1.5rem;flex-wrap:wrap;">
                ${loan.status === 'ACTIVE' ? `
                    <button class="btn btn-success btn-sm" onclick="closeModal('generic-modal'); showPaymentForm(${loan.id}, ${loan.outstanding_balance}, ${loan.interest_rate});">Registrar Pago</button>
                    <button class="btn btn-danger btn-sm" onclick="changeLoanStatus(${loan.id}, 'DELINQUENT')">Marcar en Mora</button>
                ` : ''}
                ${loan.status === 'DELINQUENT' ? `
                    <button class="btn btn-success btn-sm" onclick="closeModal('generic-modal'); showPaymentForm(${loan.id}, ${loan.outstanding_balance}, ${loan.interest_rate});">Registrar Pago</button>
                    <button class="btn btn-outline btn-sm" onclick="changeLoanStatus(${loan.id}, 'ACTIVE')">Reactivar</button>
                ` : ''}
            </div>

            <h3 style="font-size:1rem;margin-bottom:0.75rem;">Tabla de Amortización</h3>
            <div class="table-container" style="max-height:250px;overflow-y:auto;">
                <table>
                    <thead><tr><th>#</th><th>Cuota</th><th>Capital</th><th>Interés</th><th>Saldo</th></tr></thead>
                    <tbody>
                        ${loan.amortization_table.map(r => `
                            <tr>
                                <td>${r.period}</td>
                                <td class="money">${formatMoney(r.payment)}</td>
                                <td class="money">${formatMoney(r.principal_portion)}</td>
                                <td class="money">${formatMoney(r.interest_portion)}</td>
                                <td class="money">${formatMoney(r.balance)}</td>
                            </tr>
                        `).join('')}
                    </tbody>
                </table>
            </div>

            ${loan.payments.length > 0 ? `
            <h3 style="font-size:1rem;margin:1.5rem 0 0.75rem;">Historial de Pagos (${loan.payments.length})</h3>
            <div class="table-container" style="max-height:200px;overflow-y:auto;">
                <table>
                    <thead><tr><th>Fecha</th><th>Monto</th><th>Capital</th><th>Interés</th><th>Saldo</th><th>Recibo</th></tr></thead>
                    <tbody>
                        ${loan.payments.map(p => `
                            <tr>
                                <td>${formatDate(p.payment_date)}</td>
                                <td class="money">${formatMoney(p.amount)}</td>
                                <td class="money">${formatMoney(p.principal_portion)}</td>
                                <td class="money">${formatMoney(p.interest_portion)}</td>
                                <td class="money">${formatMoney(p.balance_after)}</td>
                                <td><a href="#" onclick="event.preventDefault(); showReceipt(${p.id})">${p.receipt_number}</a></td>
                            </tr>
                        `).join('')}
                    </tbody>
                </table>
            </div>` : ''}
        `);
    } catch (e) {
        showToast(e.message, 'error');
    }
}

async function changeLoanStatus(id, status) {
    try {
        await API.updateLoanStatus(id, status);
        showToast('Estado actualizado', 'success');
        closeModal('generic-modal');
        if (document.getElementById('loans-table-body')) loadLoans();
    } catch (e) {
        showToast(e.message, 'error');
    }
}

/* ── Simulator ─────────────────────────────────── */

async function renderSimulator() {
    const main = document.getElementById('main-content');
    main.innerHTML = `
        <div class="fade-in-up">
            <div class="page-header">
                <h1>🧮 Simulador de Préstamos</h1>
            </div>
            <div class="simulator-grid">
                <div class="card">
                    <h3 style="margin-bottom:1.25rem;font-size:1rem;">Parámetros del Préstamo</h3>
                    <div id="sim-form">
                        <div class="form-group">
                            <label>Capital ($)</label>
                            <input type="number" id="sim-principal" min="1" step="0.01" value="10000" placeholder="Monto del préstamo">
                        </div>
                        <div class="form-group">
                            <label>Tasa de Interés (% por periodo)</label>
                            <input type="number" id="sim-rate" min="0.01" step="0.01" value="5" placeholder="Tasa por periodo">
                        </div>
                        <div class="form-group">
                            <label>Frecuencia de Cobro</label>
                            <select id="sim-frequency">
                                <option value="MONTHLY">Mensual</option>
                                <option value="WEEKLY">Semanal</option>
                                <option value="DAILY">Diario</option>
                            </select>
                        </div>
                        <div class="form-group">
                            <label>Número de Periodos</label>
                            <input type="number" id="sim-periods" min="1" value="12" placeholder="Periodos">
                        </div>
                        <button type="button" class="btn btn-primary btn-full" id="sim-submit">Calcular</button>
                    </div>
                </div>
                <div id="sim-results">
                    <div class="card" style="text-align:center;padding:3rem;">
                        <p style="color:var(--text-muted);">Configura los parámetros y presiona "Calcular" para ver la proyección.</p>
                    </div>
                </div>
            </div>
        </div>
    `;

    document.getElementById('sim-submit').addEventListener('click', runSimulation);
}

async function runSimulation() {
    const data = {
        principal: parseFloat(document.getElementById('sim-principal').value),
        interest_rate: parseFloat(document.getElementById('sim-rate').value),
        interest_type: 'FIXED',
        payment_frequency: document.getElementById('sim-frequency').value,
        total_periods: parseInt(document.getElementById('sim-periods').value),
    };

    try {
        const result = await API.simulateLoan(data);
        const freqMap = { 'DAILY': 'Diario', 'WEEKLY': 'Semanal', 'MONTHLY': 'Mensual' };

        document.getElementById('sim-results').innerHTML = `
            <div class="sim-result fade-in-up">
                <div style="display:grid;grid-template-columns:1fr 1fr 1fr;gap:0.75rem;">
                    <div class="sim-result-card">
                        <div class="sim-result-label">Cuota ${freqMap[result.payment_frequency]}</div>
                        <div class="sim-result-value" style="color:var(--primary-light);">${formatMoney(result.periodic_payment)}</div>
                    </div>
                    <div class="sim-result-card">
                        <div class="sim-result-label">Total a Pagar</div>
                        <div class="sim-result-value" style="color:var(--warning);">${formatMoney(result.total_amount)}</div>
                    </div>
                    <div class="sim-result-card">
                        <div class="sim-result-label">Total Intereses</div>
                        <div class="sim-result-value" style="color:var(--danger);">${formatMoney(result.total_interest)}</div>
                    </div>
                </div>
                <div class="card" style="margin-top:0.5rem;">
                    <h3 style="margin-bottom:0.75rem;font-size:1rem;">Tabla de Amortización</h3>
                    <div class="table-container" style="max-height:400px;overflow-y:auto;">
                        <table>
                            <thead>
                                <tr><th>#</th><th>Cuota</th><th>Capital</th><th>Interés</th><th>Saldo</th></tr>
                            </thead>
                            <tbody>
                                ${result.table.map(r => `
                                    <tr>
                                        <td>${r.period}</td>
                                        <td class="money">${formatMoney(r.payment)}</td>
                                        <td class="money">${formatMoney(r.principal_portion)}</td>
                                        <td class="money">${formatMoney(r.interest_portion)}</td>
                                        <td class="money">${formatMoney(r.balance)}</td>
                                    </tr>
                                `).join('')}
                            </tbody>
                        </table>
                    </div>
                </div>
            </div>
        `;
    } catch (e) {
        showToast('Error en simulación: ' + e.message, 'error');
    }
}
